### PR TITLE
Various Kudu cleanups and fixes related to schema emulation races

### PR DESCRIPTION
**Pull request by grantatspothero**
## Description
Fixes lots of different problems with flaky kudu tests. Let me know if I should split some of these commits into separate PRs 
- Fixes a problem with the connector code itself racing to create the `$schemas` table
- Fixes problems with tests asserting on the number of tables that exist, which races with the `$schemas` table
- Workaround a change in the kudu client introduced in this commit that caused opening a table immediately after creating it to sometimes fail: https://github.com/apache/kudu/commit/4bf7e457b054d9bfb002e3971113e2152b92951a
- Flush kudu operations in batches instead of one at a time, this helps reduce CI runtime. This PR https://github.com/trinodb/trino/pull/10940/files changed the behavior from `AUTO_FLUSH_BACKGROUND` to `AUTO_FLUSH_SYNC` to improve correctness, but unfortunately it caused a performance degradation due to flushing each operation one at a time. This commit keeps correctness while getting better write performance.

## Related issues, pull requests, and links
https://github.com/trinodb/trino/issues/11203

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Kudu
* Fixes race in creating schemas table if schema emulation is enabled. ({issue}`11264`)
* Improve write performance by flushing operations in batches. ({issue}`11264`)
```

### CODE REVIEW

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduConnectorTest.java
Comment (grantatspothero) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduConnectorTest.java:64]: Not strictly necessary, but helps avoid test noise if the kuduServer fails to startup and then we get a bunch of NPE

File: .github/workflows/ci.yml
Comment (grantatspothero) [inline .github/workflows/ci.yml:378]: Will revert before merging.

Comment (hashhar) [reply .github/workflows/ci.yml:378]: reminder. note you do like `- { modules: plugin/trino-kudu, i: 1 }` and so on to get GHA not rate-limit you because it appears same job is being submitted multiple times.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithDisabledInferSchemaConnectorTest.java
Comment (hashhar) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithDisabledInferSchemaConnectorTest.java:41]: The comment left me confused for a while. Would it be better to say `Test methods may run in parallel and create tables in the default schema`?

Also you can use `REQUIRED_TPCH_TABLES` here to make the assertion at-least a bit stronger (since they must always exist since they get created during QueryRunnerCreation).

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java
Comment (hashhar) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:105]: Make link clickable in commit message by adding a `https://` in front.

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (hashhar) [inline plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:42]: Do we know how much heap usage growth we see when buferring all 1000 operations? i.e. is 1000 a very big value?

Can we add a configuration for this as a safety switch?

Or is not large enough to care about it at all?

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (hashhar) [inline plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:89]: See https://github.com/trinodb/trino/wiki/Error-Codes.

At some point we should define these for Kudu as well, I found them very useful as a cluster admin in the past.

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (hashhar) [inline plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:105]: nit: Maybe call these `FailedOperation`. It's fields already convey that it's OperationResponseWithError to me.

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (hashhar) [inline plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:44]: Should this be an Atomic or should we synchronize on this?

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (hashhar) [inline plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:67]: Can multiple threads call this concurrently?

If so what happens to consistency if one batch of operations fails and the other does not?

(might be a stupid question).

**Review by hashhar — COMMENTED** [review]
Thank you, lots of improvements here.

Some comments. nits are really nits, feel free to ignore.

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:42]: The default buffer size is 1000: https://github.com/apache/kudu/blob/64619147e7932518bb88b3f3f719d9c27cfdae68/java/kudu-client/src/main/java/org/apache/kudu/client/AsyncKuduSession.java#L127

So I assumed that was a reasonable default/would not consume too much memory

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:44]: Each KuduOperationApplier should only be used by a single thread, I can use `@NotThreadSafe` if we want to make that explicit

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:67]: KuduSession is not thread safe: https://kudu.apache.org/apidocs/org/apache/kudu/client/KuduSession.html

See the comment above, the assumption is this should never be called concurrently.

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:89]: Oh neat, I did not know connector specific error codes are a thing.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:105]: I had that originally and the link was large than the commit max line length 😅

File: .github/workflows/ci.yml
Comment (grantatspothero) [reply .github/workflows/ci.yml]: Had to modify `build-matrix-from-impacted.py` to support this because the git incremental builder CI changes

Comment (hashhar) [reply .github/workflows/ci.yml]: cc: @nineinchnick

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java
Comment (hashhar) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java:105]: You can refer to them like `apache/kudu@<hash>` too I believe. Or abbreviate the commit hash until it fits (any length works as long as the hash is not ambiguous).

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (hashhar) [reply plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:45]: Perfect.

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (hashhar) [reply plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:47]: cc: @findepi Do we have convention? I see some classes marked `@ThreadSafe` though. Do we assume not thread safe if not annotated?

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (hashhar) [reply plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:70]: Do mention this in the class level and this method's javadoc (if we don't end up adding @NotThreadSafe).

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (findepi) [reply plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:47]: > Do we assume not thread safe if not annotated?

more or less (ie unless it's clear the class needs to be thread safe

we don't have a convention to annotate `@NotThreadSafe` those who're not.

Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:47]: I'll remove the annotation and add to the javadoc then 👍

File: .github/bin/build-matrix-from-impacted.py
Comment (hashhar) [inline .github/bin/build-matrix-from-impacted.py:73]: Let's do this separately. Will need more work to allow arbitrary keys instead of hardcoded ones.

Comment (nineinchnick) [reply .github/bin/build-matrix-from-impacted.py:73]: I'll open up another pr

File: .github/workflows/ci.yml
Comment (nineinchnick) [reply .github/workflows/ci.yml]: https://github.com/trinodb/trino/pull/11446

File: .github/bin/build-matrix-from-impacted.py
Comment (nineinchnick) [reply .github/bin/build-matrix-from-impacted.py:73]: https://github.com/trinodb/trino/pull/11446

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (Praveen2112) [inline plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:50]: Can we configure the `kuduSession` to have manual_flush as its default flush mode.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithStandardInferSchemaConnectorTest.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithStandardInferSchemaConnectorTest.java:43]: "if the name is fixed" can you explain what you mean? I'm not following.

Comment (hashhar) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithStandardInferSchemaConnectorTest.java:43]: I believe that since the table we exclude is always called "$schema" we can just make it part of query like `SHOW TABLES FROM default NOT LIKE '$schema'` or maybe use one of the assertions like `assertThat(row).containsExactlyInAnyOrderElementsOf(<row created by hand>)`. (search for `containsExactlyInAnyOrder` to see examples).

File: plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java:50]: It is tricky to do without introducing extra abstraction in `ForwardingKuduClient`:
https://github.com/trinodb/trino/blob/master/plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/ForwardingKuduClient.java#L98-L101

And we guard against the case of incorrectly using the wrong flush mode here. I'm inclined to leave the configured default flush mode.

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java
Comment (leveyja) [inline plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java:117]: Q: Should `nextSubId` be marked `volatile` (it's a member being updated in a CompletableFuture)?

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/schema/SchemaEmulationByTableNameConvention.java
Comment (leveyja) [inline plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/schema/SchemaEmulationByTableNameConvention.java:174]: Q: (For complete correctness) should we do a comparison on the existing table's options to be sure they're effectively "the same"?
(I know this is fixing a race in test code, but in Prod code if someone attempted to create a table w/incompatible options at the same time as someone else, they may (unexpectedly) get back a KuduTable which doesn't match the create options they were trying to apply?)

Comment (leveyja) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/schema/SchemaEmulationByTableNameConvention.java:174]: (Looking a little more closely, it seems the only usage is from `createAndFillSchemasTable`, which means the case I suggested can't manifest - so ignore!).

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java:117]: The completablefuture `NOT_BLOCKED` is just a placeholder to signal the page sink is not blocked. `NOT_BLOCKED` does not hold a reference to `nextSubId`, so I do not think there are any issues here.

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/schema/SchemaEmulationByTableNameConvention.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/schema/SchemaEmulationByTableNameConvention.java:174]: The whole approach of emulating schemas in kudu using a kudu table is not safe given[ kudu's consistency semantics ](https://kudu.apache.org/docs/transaction_semantics.html) and trino does not propagate client timestamps in kudu.

I talked with some kudu maintainers, impala uses the hive metastore to emulate schemas in kudu, that introduces a big dependency however.

File: plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java
Comment (leveyja) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java:117]: Just to be clear, non-volatile means that any other thread may not see the new value of `nextSubId++` (it's not volatile, so there's no guarantee a separate thread will ever see the updated value, if it had ever read the value before). 
If that's not important, then feel free to ignore! (But as an aside, I think it's generally better to be explicit about these scenarios, and I would mark it volatile to be correct - rather than have code which could be "incorrect" under certain circumstances, and just depending on those situations being "impossible" - someone may always re-use this code and fall foul of the invisible update!).

Comment (leveyja) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java:117]: Concrete example:
```
executor.submit(() -> kuduPageSink.appendPage(page));
executor.submit(() -> kuduPageSink.appendPage(anotherPage));
```
If that were a multi-threaded Exector, this line could yield the same value in both threads:
```
String id = format("%s-%08x", uuid, nextSubId++);
```
(both threads read kuduPageSink.nextSubId , both increment it, neither see the value of the other, so it's both a race and a lost update - but even if only 1 thread did updates, because it's not volatile, there's no guarantee anyone who has a reference to kuduPageSink instance will see the update `nextSubId++`, hence the problem).

On reflection, I'd use `AtomicInteger.getAndIncrement()` for simplicity sake (and to avoid the race I mentioned - volatile wouldn't protect against that...).

But anyways - just wanted to throw that out there for your consideration!

Comment (grantatspothero) [reply plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduPageSink.java:117]: Correct if I'm wrong, but page sinks are not expected to be thread safe. 

No where in the javadoc for `ConnectorPageSink` does it list thread safety as a requirement. The closest thing I could find was:

```
    /**
     * Returns a future that will be completed when the page sink can accept
     * more pages.  If the page sink can accept more pages immediately,
     * this method should return {@code NOT_BLOCKED}.
     */
    CompletableFuture<?> appendPage(Page page);
``` 

Here we only return `NOT_BLOCKED` after the critical section, so there is no risk of concurrent access to `nextSubId`

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithStandardInferSchemaConnectorTest.java
Comment (Praveen2112) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithStandardInferSchemaConnectorTest.java:35]: > I think the purpose of the table is listing every table from every tablet is expensive, so instead do that listing once and cache the results in a kudu table and then update the kudu table whenever a new schema is created/deleted.

I think we don't use it for caching - Kudu doesn't have the concept of namespace i.e schema so we virtually create a table to store the schemas and add them as a prefix for the tables under this virtual schema. 
https://trino.io/docs/current/connector/kudu.html#behavior-with-schema-emulation

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithDisabledInferSchemaConnectorTest.java
Comment (Praveen2112) [inline plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithDisabledInferSchemaConnectorTest.java:49]: nit: Any specific reason on not using similar `toUnmodifiableList`here ?

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithDisabledInferSchemaConnectorTest.java
Comment (Praveen2112) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithDisabledInferSchemaConnectorTest.java:42]: Any reason on wrapping it on top of `ArrayList` instead of using `computeActual("SHOW TABLES FROM default").getMaterializedRows().stream()`

**Review by Praveen2112 — APPROVED** [review]

**Comment by grantatspothero** [comment]
@Praveen2112 WRT to https://github.com/trinodb/trino/pull/11264#discussion_r832923091

Yes, but you could imagine instead of creating a $schemas table to emulate schemas you could list all kudu tables and fro the naming convention use that information to emulate schemas. The $schemas table is purely for performance reasons.

**Comment by Praveen2112** [comment]
Yeah !!  I agree schema operations would be a lot faster with the $schema tables.

File: plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithDisabledInferSchemaConnectorTest.java
Comment (grantatspothero) [reply plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/AbstractKuduWithDisabledInferSchemaConnectorTest.java:49]: Doesn't seem like it matters given the list clearly cannot be modified.

**Review by hashhar — APPROVED** [review]

**Comment by hashhar** [comment]
Thanks a lot for digging to the bottom of this flakiness @grantatspothero . Wasn't easy for sure. And the performance improvements for writes is also nice.

**Pull request merged**
